### PR TITLE
#322 Fluent syntax for applying controlled gates

### PR DESCRIPTION
**Pull request by ptrc-n**
CNOT Gate can now optionally be called with keyword arguments 'target' and 'control'.

### CODE REVIEW

File: cirq/ops/common_gates.py [hunk 1]
```diff
@@ -505,7 +505,8 @@ class CNotGate(eigen_gate.EigenGate,
                gate_features.CompositeGate,
                gate_features.TwoQubitGate,
                gate_features.QasmConvertableGate):
-    """A controlled-NOT. Toggle the second qubit when the first qubit is on."""
+    """A controlled-NOT. Toggle the second qubit when the first qubit is on.
+    You can also use the keyword arguments 'target' and 'control' instead of the positional ones."""
```
Comment (Strilanc) [inline cirq/ops/common_gates.py]: This sentence is a bit out of place. I would go with something more like

"When applying CNOT to qubits, you can either use positional arguments `CNOT(q1, q2)`  or named arguments `CNOT(control=q1, target=q2)`. (Mixing the two is not permitted.)"

File: cirq/ops/common_gates.py [hunk 2]
```diff
@@ -575,6 +576,11 @@ def __repr__(self) -> str:
             return 'CNOT'
         return 'CNOT**{!r}'.format(self.half_turns)
 
+    def on(self, *args: raw_types.QubitId, **kwargs: raw_types.QubitId) -> gate_operation.GateOperation:
+        if 'control' in kwargs.keys() and 'target' in kwargs.keys():
```
Comment (Strilanc) [inline cirq/ops/common_gates.py]: Make sure to fail when a mix of named and positional arguments are specified:

```python
if args and kwargs:
    raise ValueError('Mixed named/positional control/target.')
```

Should also fail when there are unknown kwargs.

File: cirq/ops/common_gates.py [hunk 3]
```diff
@@ -575,6 +576,11 @@ def __repr__(self) -> str:
...
+            return super(eigen_gate.EigenGate, self).on(kwargs['control'], kwargs['target'])
```
Comment (Strilanc) [inline cirq/ops/common_gates.py]: just `super()`

File: cirq/ops/common_gates.py [hunk 4]
```diff
@@ -575,6 +576,11 @@ def __repr__(self) -> str:
...
+        return super(eigen_gate.EigenGate, self).on(*args)
```
Comment (Strilanc) [inline cirq/ops/common_gates.py]: just `super()`

File: cirq/ops/common_gates_test.py [hunk 5]
```diff
@@ -256,6 +256,22 @@ def test_cnot_power():
...
+    a = cirq.NamedQubit('a')
+    b = cirq.NamedQubit('b')
+
+    circuit_positional = cirq.Circuit.from_ops(
+        cirq.CNOT(a,b),
+        cirq.CNOT(b, a)
+    )
+
+    circuit_keywords = cirq.Circuit.from_ops(
+        cirq.CNOT(control=a, target=b),
```
Comment (Strilanc) [inline cirq/ops/common_gates_test.py]: No need to make a circuit, just use `cirq.testing.EqualsTester` and drop `CNOT(a, b)` and `CNOT(control=a, target)` into one group and the other two into another equivalence group.

File: cirq/ops/common_gates_test.py [hunk 6]
```diff
@@ -256,6 +256,22 @@ def test_cnot_power():
...
+        cirq.CNOT(target=a, control=b)
+    )
+
+    assert circuit_positional == circuit_keywords
+
```
Comment (Strilanc) [inline cirq/ops/common_gates_test.py]: Add a test of the failure cases (e.g. `CNOT(b, target=a)` should raise a `ValueError`).

**Comment by Strilanc** [comment]
Also, if you go into the travis CI logs (click on "details" in the "All checks have failed" section) you will find several lint warnings:

https://travis-ci.com/quantumlib/Cirq/jobs/136844767

```
C:509, 0: Line too long (100/80) (line-too-long)
C:509, 0: Line too long (100/80) (line-too-long)
C:579, 0: Line too long (104/80) (line-too-long)
C:581, 0: Line too long (92/80) (line-too-long)
E:581,19: Bad first argument 'EigenGate' given to super() (bad-super-call)
E:582,15: Bad first argument 'EigenGate' given to super() (bad-super-call)
```

File: cirq/ops/common_gates.py [hunk 7]
```diff
@@ -575,6 +578,28 @@ def __repr__(self) -> str:
             return 'CNOT'
         return 'CNOT**{!r}'.format(self.half_turns)
 
+    def on(self, *args: raw_types.QubitId,
+           **kwargs: raw_types.QubitId) -> gate_operation.GateOperation:
+        if kwargs and args:
+            raise ValueError('A mix of named and positional arguments has been '
+                             'specified for a CNOT operation.\n'
+                             'Be sure to choose either one of them')
+        for argkey in kwargs.keys():
```
Comment (Strilanc) [inline cirq/ops/common_gates.py]: This loop and the following two conditions can be simplified into this:

```python
if set(kwargs.keys()) not in [{}, {'control', 'target'}]:
    raise ValueError("Expected 'control' and 'target' keyword arguments, or no keyword arguments, but got {}.".format(set(kwargs.keys()))
```

Or.. I guess the method could be written like so

```python
if not kwargs:
    return super().on(*args)
if not args and set(kwargs.keys) == {'control', 'target'}:
    return super().on(kwargs['control'], kwargs['target'])
raise ValueError("Expected two positional argument or else 'target' and 'control' keyword arguments. But got args={!r}, kwargs={!r}.".format(args, kwargs))
```

Comment (ptrc-n) [reply cirq/ops/common_gates.py]: I tried to find a good balance between readability and lines of code.
Thought the solution I came up with would be okey already, but I liked your suggestion better.

**Pull request merged**